File: elixir/palindrome-products/palindrome_products.exs
defmodule Palindromes do

  @doc """
  Generates all palindrome products from an optionally given min factor (or 1) to a given max factor.
  """
  @spec generate(non_neg_integer, non_neg_integer) :: map
  def generate(max_factor, min_factor \\ 1) do
    for a <- min_factor..max_factor,
      b <- a..max_factor,
      palindrome?(a*b) do
      [a, b]
    end
    |> Enum.reduce(%{}, fn [a, b], acc ->
      Map.update(acc, a*b, [[a, b]], &(&1 ++ [[a, b]]))
    end)
  end

  @spec palindrome?(integer) :: boolean
  def palindrome?(n) do
    digits = Integer.digits(n)
    Enum.reverse(digits) == digits
  end
end

File: python/palindrome-products/palindrome_products.py
def largest_palindrome(min_factor, max_factor):
    if min_factor > max_factor:
        raise ValueError('Invalid factors.')

    # reduce the search space
    if max_factor < 10:
        factors = [(a, b) for a in range(max_factor, min_factor-1, -1)
                   for b in range(a, min_factor-1, -1)
                   if is_palindrome(a * b)]
        palindrome = max(a * b for (a, b) in factors)
        factors = {(a, b) for (a, b) in factors if a * b == palindrome}
        return (palindrome, factors)

    min_factor = int(max_factor * 0.8) if max_factor < 100 else int(max_factor * 0.9)

    for a in range(max_factor, min_factor, -1):
        for b in range(a, min_factor, -1):
            if is_palindrome(a * b):
                return (a * b, {(a, b)})

    raise ValueError('No palindrome products in range.')


def smallest_palindrome(min_factor, max_factor):
    if min_factor > max_factor:
        raise ValueError('Invalid factors.')

    for a in range(min_factor, max_factor-1):
        for b in range(a, max_factor-1):
            if is_palindrome(a * b):
                return (a * b, {(a, b)})

    raise ValueError('No palindrome products in range.')


def is_palindrome(n):
    return str(n)[::-1] == str(n)
